### PR TITLE
add print_query_id support for native client

### PR DESCRIPTION
**Pull request by amosbird**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Add a new option `print_query_id` to clickhouse-client. It helps generate arbitrary strings with the current query id generated by the client.

Detailed description / Documentation draft:

example use case:

config-client.xml
```
<config>
  <query_id_formats>
    <speedscope>http://speedscope-host/#profileURL=qp%3Fid%3D{query_id}</speedscope>
  </query_id_formats>
</config>
```

`clickhouse-client --print_query_id --config config-client.xml`

Execute some heavy query:
```
:) insert into function file('/tmp/cd', 'Parquet', 'i int') select number from numbers(100000000)

INSERT INTO FUNCTION file('/tmp/cd', 'Parquet', 'i int') SELECT number
FROM numbers(100000000)

speedscope:
http://speedscope-host/#profileURL=qp%3Fid%3D5182b958-bbd9-4d5e-b448-a41632359554

Ok.

0 rows in set. Elapsed: 2.939 sec. Processed 100.66 million rows, 805.28 MB (34.25 million rows/s., 274.02 MB/s.)

:)
```

The query profile url is instantly available.

ClickHouse contains extraordinary capabilities to do introspecting, however, it requires `query_id` to access related metrics. With custom http handlers and this PR, it makes perf tuning much more pleasant.

### CODE REVIEW

File: programs/client/Client.cpp [hunk 1]
```diff
@@ -2264,6 +2292,7 @@ class Client : public Poco::Util::Application
             ("ask-password", "ask-password")
             ("quota_key", po::value<std::string>(), "A string to differentiate quotas when the user have keyed quotas configured on server")
             ("query_id", po::value<std::string>(), "query_id")
+            ("print_query_id", "Generate UUID query id for each query and print it")
```
Comment (alexey-milovidov) [inline programs/client/Client.cpp:2295]: I think, we can generate query_id by default, so we don't need this config parameter.

File: src/Interpreters/Context.cpp [hunk 2]
```diff
@@ -1109,43 +1109,43 @@ void Context::setCurrentDatabase(const String & name)
...
+    return QueryUUID(random.bytes, Poco::UUID::UUID_RANDOM).toString();
+}
 
-        /// Use protected constructor.
-        struct QueryUUID : Poco::UUID
-        {
-            QueryUUID(const char * bytes, Poco::UUID::Version version)
-                : Poco::UUID(bytes, version) {}
-        };
+void Context::setCurrentQueryId(const String & query_id)
```
Comment (alexey-milovidov) [inline src/Interpreters/Context.cpp:1140]: Client also has Context. We can simply reuse this method to set or generate query id (for both cases: where there is query_id specified explicitly in command line and for the case when we have to generate it). And then we can just obtain query_id from Context.

File: programs/client/Client.cpp [hunk 3]
```diff
@@ -243,6 +246,17 @@ class Client : public Poco::Util::Application
...
+        /// Initialize query_id_formats if any
+        if (config().has("query_id_formats"))
+        {
+            Poco::Util::AbstractConfiguration::Keys keys;
+            config().keys("query_id_formats", keys);
+            for (const auto & name : keys)
+                query_id_formats.emplace_back(name + ":\n", config().getString("query_id_formats." + name));
+        }
+        if (query_id_formats.empty())
+            query_id_formats.emplace_back("Query id:\n", "{query_id}");
```
Comment (alexey-milovidov) [inline programs/client/Client.cpp:259]: Why need \n?

**Review by alexey-milovidov — COMMENTED** [review]
Maybe add a test with `expect`?

File: programs/client/Client.cpp (hunk 1 above)
Comment (amosbird) [reply programs/client/Client.cpp:2295]: The behavior somehow changes when `print_query_id = true` as we always generate query_id on the client side.

File: programs/client/Client.cpp (hunk 3 above)
Comment (amosbird) [reply programs/client/Client.cpp:259]: URL hinters usually prefers one-line strings. Although some terminals can correctly understand line wraps, it won't work for screen or tmux. With `\n` it's more likely to generate one-line urls.

File: programs/client/Client.cpp (hunk 1 above)
Comment (alexey-milovidov) [reply programs/client/Client.cpp:2295]: That's fine if client will always generate query_id.

File: programs/client/Client.cpp (hunk 3 above)
Comment (alexey-milovidov) [reply programs/client/Client.cpp:259]: I still consider "speedscope" as minor use case in contrast to the default behaviour when client will just display something like
```
Query id: 01db8425-415a-4ec2-bcbf-14dbc49aa8dd
```
for every query.

File: src/Interpreters/Context.cpp (hunk 2 above)
Comment (amosbird) [reply src/Interpreters/Context.cpp:1140]: However, it'll break the previous assumption 
 https://github.com/ClickHouse/ClickHouse/blob/master/src/Interpreters/Context.cpp#L1115

Comment (alexey-milovidov) [reply src/Interpreters/Context.cpp]: Ok, got it.

**Comment by azat** [comment]
>ClickHouse contains extraordinary capabilities to do introspecting, however, it requires query_id to access related metrics. With custom http handlers and this PR, it makes perf tuning much more pleasant.

Interesting, when I thought about this I've come up with [`previousQueryId`](https://github.com/azat/ClickHouse/commit/9c496d9e133fc00b61baec003c56e3e2d24d48c3#diff-f8da3667d31f93af44f6b424885f44c5), but I hadn't submit it because I thought that it is unusable in tests, however now I think that it can be used:

```sql
set log_queries=1; -- let's make it explicit
select 1;
system flush logs settings log_queries=0; -- that bit that was missing that time is updating previousQueryId only if log_queries is enabled
select * from system.query_log where query_id = previousQueryId();
```

**Comment by amosbird** [comment]
> Maybe it is better to add previousQueryId + some client hooks

That sounds no better either :) .  previousQueryId is interesting but client now becomes stateful, which is bad.

**Comment by azat** [comment]
>but client now becomes stateful

Not the client, since the information is stored in the client context on the server, but agree - explicit query_id is better
Those query_id_formats looks too specific though

>system flush logs settings log_queries=0; -- that bit that was missing that time is updating previousQueryId only if log_queries is enabled

Anyway this does not work now (there was an attempt to make it work in #11804)

**Comment by amosbird** [comment]
> Those query_id_formats looks too specific though

The default is `Query id: xxx`.

The idea of `query_id_formats` is inspired from impala-client, which prints a query profiling url for every query. The url presents a nice dynamic DAG data flow, which excels our progress bar.

BTW, it's also possible to print OSC sequence so that query id is automatically stored into your clipboard, or launch something if you use a customized terminal.

**Comment by alexey-milovidov** [comment]
> Docs check — Found 1 errors

Infrastructure error.

**Comment by alexey-milovidov** [comment]
> launch something if you use a customized terminal

This will be useful for `EXPLAIN PIPELINE graph = 1`: https://presentations.clickhouse.tech/meetup46/explain.pdf
To open a window with graph viewer...

**Review by alexey-milovidov — APPROVED** [review]

**Comment by alexey-milovidov** [comment]
Not enough beautiful.

Screenshot_20201011_180707

I would like to add one extra newline after Query id: 6b355015-08fd-4e55-883f-8c9d9a93728f

**Comment by alexey-milovidov** [comment]
No, it's too ugly :(

Screenshot_20201011_181007

**Pull request merged**